airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,280 @@
+from airbyte_cdk.sources.source import TState
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.concurrent.abstract_stream import AbstractStream
+from airbyte_cdk.sources.streams.concurrent.adapters import CursorPartitionGenerator
+from airbyte_cdk.sources.streams.concurrent.availability_strategy import AlwaysAvailableAvailabilityStrategy
+from airbyte_cdk.sources.streams.concurrent.cursor import ConcurrentCursor, CursorField, CursorValueType, GapType
+from airbyte_cdk.sources.streams.concurrent.default_stream import DefaultStream
+from airbyte_cdk.sources.streams.concurrent.helpers import get_primary_key_from_stream
+from airbyte_cdk.sources.streams.concurrent.state_converters.datetime_stream_state_converter import (
+    CustomOutputFormatConcurrentStreamStateConverter,
+)
+from isodate import parse_duration
+
+
+@dataclass
+class DeclarativeCursorAttributes:
+    cursor_field: CursorField
+    datetime_format: str
+    slice_boundary_fields: Optional[Tuple[str, str]]
+    start: Optional[CursorValueType]
+    end_provider: Optional[Callable[[], CursorValueType]]
+    lookback_window: Optional[GapType]
+    slice_range: Optional[GapType]
+    cursor_granularity: Optional[GapType]
+
+
+class ConcurrentDeclarativeSource(ManifestDeclarativeSource, Generic[TState]):
+    def __init__(
+        self,
+        catalog: Optional[ConfiguredAirbyteCatalog],
+        config: Optional[Mapping[str, Any]],
+        state: TState,
+        source_config: ConnectionDefinition,
+        debug: bool = False,
+        emit_connector_builder_messages: bool = False,
+        component_factory: Optional[ModelToComponentFactory] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(
+            source_config=source_config,
+            debug=debug,
+            emit_connector_builder_messages=emit_connector_builder_messages,
+            component_factory=component_factory,
+        )
+
+        self._state = state
+
+        self._concurrent_streams: List[AbstractStream]
+        self._synchronous_streams: List[Stream]
+
+        self._concurrent_streams, self._synchronous_streams = self._separate_streams(config=config or {})
+
+        concurrency_level_from_manifest = self._source_config.get("concurrency_level")
+        if concurrency_level_from_manifest:
+            concurrency_level_component = self._constructor.create_component(
+                model_type=ConcurrencyLevelModel, component_definition=concurrency_level_from_manifest, config=config or {}
+            )
+            if not isinstance(concurrency_level_component, ConcurrencyLevel):
+                raise ValueError(f"Expected to generate a ConcurrencyLevel component, but received {concurrency_level_component.__class__}")
+
+            concurrency_level = concurrency_level_component.get_concurrency_level()
+            initial_number_of_partitions_to_generate = concurrency_level // 2
+        else:
+            concurrency_level = 1
```

Comment:
I think it would be better to move the value from the code into a class variable may be to improve readability